src/control/security/auth/auth_sys.go

```diff
@@ -66,239 +70,83 @@ func VerifyToken(key crypto.PublicKey, token *Token, sig []byte) error {
-	if err != nil {
-		return nil, errors.WithMessage(err, "Unable to generate verifier")
+type (
+	AuthMap map[Flavor]CredentialRequestFactory
+
+	CredentialRequestFactory interface {
+		// Allocate, and return, an instance of the type that implements CredentialRequest interface.
+		AllocCredentialRequest() CredentialRequest
+		// Returns a unique 4 character identifier (AuthTag) that refers to the CredentialRequest implementation.
+		GetAuthFlavor() Flavor
+	}
+
+	CredentialRequest interface {
+		// Using information stored in the agent's security config, the session/socket, request body, and the agent's signing key,
+		// initalize the state of the Credential Request so that it can sign a credential in the future. Return an error if
+		// a problem occurs during initializiation.
+		InitCredentialRequest(log logging.Logger, sec_cfg *security.CredentialConfig, session *drpc.Session, req_body []byte, key crypto.PrivateKey) error
+		// Contact a source of authenticity (e.g. domain socket, access manager) using the initalized state of the CredentialRequest
+		// to construct a Credential object.
+		GetSignedCredential(log logging.Logger, ctx context.Context) (*Credential, error)
+		// Returns a key, as a string, representing a unique identifer specific to the request. This key is used by the cache
+		// to remember credentials. It is vital for security that this identifer cannot be forged or easily guessed by the client - 
+		// otherwise cached credentials can be "stolen".
+		CredReqKey() string
 	}
+)
 
-	verifierToken := Token{
-		Flavor: Flavor_AUTH_SYS,
-		Data:   verifier}
-
-	credential := Credential{
-		Token:    &token,
-		Verifier: &verifierToken,
-		Origin:   "agent"}
-
-	logging.FromContext(ctx).Tracef("%s: successfully signed credential", req.DomainInfo)
-	return &credential, nil
-}
+// CredentialRequests is a list of authentication methods the agent can use.
+// To implement a new type of authentication: satisfy the CredentialRequest and 
+// CredentialRequestFactory interfaces, add a new flavor in auth.proto, ensure
+// that your `GetAuthFlavor` method returns this new unique flavor and add your 
+// interface to the `CredentialRequests` list below.
+// The server must be configured to allow an authentication method when it is initalized.
+// By default, only Unix authentication is enabled.
 
-// AuthSysFromAuthToken takes an opaque AuthToken and turns it into a
-// concrete AuthSys data structure.
-func AuthSysFromAuthToken(authToken *Token) (*Sys, error) {
-	if authToken.GetFlavor() != Flavor_AUTH_SYS {
-		return nil, errors.New("Attempting to convert an invalid AuthSys Token")
-	}
-
-	sysToken := &Sys{}
-	err := proto.Unmarshal(authToken.GetData(), sysToken)
-	if err != nil {
-		return nil, errors.Wrapf(err, "unmarshaling %s", authToken.GetFlavor())
-	}
-	return sysToken, nil
-}
+var CredentialRequests = []CredentialRequestFactory{&CredentialRequestUnix{}, &CredentialRequestAM{}}
```

Comment:
Revisit why I implemented it this way. Can this be simplified? Maybe just a function with a mapping/switch statement between flavors and factories.

Reply:
Split the factory and the request into two separate things. The requests should not double as factories; the factories should exist only to generate requests!